[API] Null-Link handling

NobodysNightmare opened the pull request:
## OpenProject Work Package

https://community.openproject.org/work_packages/17894

NobodysNightmare committed:
change links to nothing in work package

- now a link exists that has a null href
-> instead of hiding the whole link
- refactored generation of links to be more uniform

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-null-links.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-null-links.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
@ulferts It would be great if you could have an early look on this PR. I'd like to know what you like/dislike about the new `linked_property` method, because I want to introduce it in all representers.

NobodysNightmare committed:
also use new link helper in version representer

NobodysNightmare committed:
extract helper for self links

NobodysNightmare commented on lib/api/v3/projects/project_representer.rb:
```diff
@@ -34,12 +34,8 @@ module API
-            title: "#{represented.name}"
-          }
-        end
+
```
I personally like those extra empty lines, Sir!

@ulferts @myabc What do you think?

myabc replied:
@NobodysNightmare I agree with you. I don't mind an extra line here. Hound/rubocop seems to have got stricter about empty lines of late. Perhaps we can disable this cop?

NobodysNightmare replied:
``` yml
EmptyLinesAroundClassBody:
  Enabled: false
```

Should I?

NobodysNightmare committed:
remove unneccessary selfs

NobodysNightmare committed:
fix integration tests

NobodysNightmare committed:
just don't say anything hound!

NobodysNightmare commented:
@ulferts I'd say this is ready now... mind my remarks at the @houndci comment...

NobodysNightmare committed:
rename visible_condition to show_if

NobodysNightmare commented:
I'll invest some more work before declaring this finished...

NobodysNightmare committed:
also apply visibility conditions to parent

- parent link only invisible if present and not visible to user
- no parent rendered as null link
- tests for all three cases (none, visible, invisible)

NobodysNightmare commented:
@ulferts ready for you to review ^^ (or anyone else)

NobodysNightmare committed:
fix parameter indentation

ulferts commented:
@NobodysNightmare now this needs rebasing

NobodysNightmare committed:
fix and expand category specs

NobodysNightmare commented:
@ulferts rebasing complete... lets wait for travis I guess ^^

edit: Travis is **green** as one can not see....

ulferts commented on lib/api/decorators/single.rb:
```diff
@@ -51,6 +51,29 @@ def initialize(model, context = {})
+
+          link_object
+        end
+      end
```
can't we even make a sensible default here to use the first part of the representer name if nothing is provided so that one can just write:

```
self_link
```

ulferts replied:
Or we can use the `type` attribute of the decorator. That should be about the same and a little less magical.

ulferts commented:
@NobodysNightmare because I just noticed the children property I am wondering what the specification is with regards to a list of linked properties. Currently, the field will be empty.

ulferts commented on lib/api/decorators/single.rb:
```diff
@@ -51,6 +51,29 @@ def initialize(model, context = {})
+
+      def self.linked_property(property,
+                               path: property,
+                               backing_field: property,
```
In Rails terms, this would be an association. From what I have seen, I think that the method is limited to representing associations, so renaming it to such would make sense. 
By the way, if backing_field is not set, you could also take property as a default. Same is true for path.

NobodysNightmare replied:
Am I missing something or **do** I already set property as default for both?

So you propose `association` as name, right?

NobodysNightmare committed:
make self_link even more awesome

NobodysNightmare committed:
use show_if as guard clause

NobodysNightmare committed:
rename parameter

NobodysNightmare commented:
if you look closely, you will see that the travis build **succeeded**!
